opnode/rollup/sync/start.go

```diff
@@ -9,102 +9,140 @@ import (
+	// Starting point
+	l2Head, err := source.L2NodeByNumber(ctx, nil, genesis)
+	if err != nil {
+		return L2Node{}, fmt.Errorf("failed to fetch L2 head: %w", err)
 	}
-	if currentL1 == refL1 {
-		// L1 node has head-block of execution-engine, so we should fetch the L1 block that builds on top.
-		var ontoL1 eth.BlockID // ontoL1 is the parent, to make sure we got a nextRefL1 that connects as expected.
-		nextRefL1, ontoL1, err = reference.RefByL1Num(ctx, refL1.Number+1)
+	// Walk L2 chain from L2 head to first L2 block which has a L1 Parent that is canonical. May walk to L2 genesis
+	for n := l2Head; ; {
+		l1header, err := source.L1NodeByNumber(ctx, n.l1parent.Number)
 		if err != nil {
-			// If refL1 is the head block, then we might not have a next block to build on the head
-			if errors.Is(err, ethereum.NotFound) {
-				// return the same as the engine head was already built on, no error.
-				nextRefL1 = refL1
-				refL2 = eth.BlockID{Hash: parentL2, Number: refL2.Number}
-				if refL2.Number > 0 {
-					refL2.Number -= 1
-				}
-				err = nil
-				return
+			// Generic error, bail out.
+			if !errors.Is(err, ethereum.NotFound) {
+				return L2Node{}, fmt.Errorf("failed to fetch L1 block %v: %w", n.l1parent.Number, err)
+			}
+			// L1 block not found, keep walking chain
+		} else {
+			// L1 Block found, check if matches & should keep walking the chain
+			if l1header.self.Hash == n.l1parent.Hash {
+				return n, nil
 			}
-			return
 		}
-		// The L1 source might rug us with a reorg between API calls, catch that.
-		if ontoL1 != currentL1 {
-			err = fmt.Errorf("the L1 source reorged, the block for N+1 %s doesn't have the previously fetched block N %s as parent, but builds on %s instead", nextRefL1, currentL1, ontoL1)
+
+		// Don't walk past genesis. If we were at the L2 genesis, but could not find the L1 genesis
+		// pointed to from it, we are on the wrong L1 chain.
+		if n.self.Hash == genesis.L2.Hash || n.self.Number == genesis.L2.Number {
+			return L2Node{}, WrongChainErr
 		}
-		return
-	}
 
-	// Search back: linear walk back from engine head. Should only be as deep as the reorg.
-	for refL2.Number > 0 {
-		// remember the canonical L1 block that builds on top of the L1 source block of the L2 parent block.
-		nextRefL1 = currentL1
-		refL1, refL2, parentL2, err = reference.RefByL2Hash(ctx, parentL2, genesis)
+		// Pull L2 parent for next iteration
+		n, err = source.L2NodeByHash(ctx, n.l2parent.Hash, genesis)
 		if err != nil {
-			// TODO: re-attempt look-up, now that we already traversed previous history?
-			err = fmt.Errorf("failed to lookup block %s in L2: %w", refL2, err) // refL2 is previous parentL2
-			return
+			return L2Node{}, fmt.Errorf("failed to fetch L2 block by hash %v: %w", n.l2parent.Hash, err)
 		}
-		// Check if L1 source has the block that derived the L2 block we are planning to build on
-		currentL1, _, err = reference.RefByL1Num(ctx, refL1.Number)
+	}
+}
```

Comment:
I disagree here: I think that the L2 chains should follow what the L1 chain does - including deep re-orgs. I don't think we have well defined semantics for what happens if we keep extending the L2 chain based on a L1 chain that is no longer canonical.

I think we could throw an error and halt the L2 chain here, but continuing to produce blocks / sync doesn't really work.

Reply:
Yeah, I agree, but that was also my point: if it reorgs past 100 L1 blocks (or whatever number > 13 minutes we pick), halt and catch fire, because it means that something is probably very wrong, and there is no point trying to do a deeper re-org than that.

So when I say "prevent re-orging" I really mean "abort because something's wrong" :)

Reply:
Well now there's a panic. It's probably worth making an unrecoverable error tho.